### PR TITLE
[fsdp, model] feat: support glm_moe_dsa FSDP training with DSA attention

### PR DESCRIPTION
**Pull request by Kite0011**
### What does this PR do?
Add FSDP training support for the glm_moe_dsa model, including:
+ Custom attention forward with DSA (Dynamic Sparse Attention) and Ulysses sequence parallelism support
+ Monkey patch registration for glm_moe_dsa model type
+ FLOPs estimation function accounting for MLA + DSA indexer + sparse attention

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [ ] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

loss converge using a random glm_moe_dsa model 10 step
task: open_reasoning_math sft
Clipboard_Screenshot_1780041072

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/verl-project/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [ ] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

File: verl/utils/flops_counter.py [hunk 1]
```diff
@@ -535,6 +535,77 @@ def _estimate_gpt_oss_flops(config, tokens_sum, batch_seqlens, delta_time):
...
+
+    # DSA indexer linear params per layer:
+    # wq_b: q_lora_rank -> index_n_heads * index_head_dim
+    # wk: kv_lora_rank -> index_n_heads * index_head_dim
+    # weights_proj: index_n_heads -> num_attention_heads
+    indexer_linear_N = (
+        q_lora_rank * (index_n_heads * index_head_dim)
+        + kv_lora_rank * (index_n_heads * index_head_dim)
+        + index_n_heads * num_attention_heads
+    )
```
[the remark this answers is not in the document]
Comment (Kite0011) [reply verl/utils/flops_counter.py:584]: yes you are right

File: verl/models/transformers/monkey_patch.py [hunk 2]
```diff
@@ -523,6 +523,15 @@ def state_dict(self, *args, **kwargs):
...
             patch_vlm_for_ulysses_input_slicing(Qwen3_5MoeTextModel)
 
+    elif model.config.model_type == "glm_moe_dsa":
+        if use_remove_padding or ulysses_sp_size > 1:
+            from transformers.models.glm_moe_dsa.modeling_glm_moe_dsa import GlmMoeDsaAttention
+
+            from verl.models.transformers.glm_moe_dsa import glm_moe_dsa_attn_forward_with_dsa
+
+            GlmMoeDsaAttention.forward = glm_moe_dsa_attn_forward_with_dsa
+            print(f"Monkey patch {model.__class__.__name__} attention layer with DSA")
```
[the remark this answers is not in the document]
Comment (Kite0011) [reply verl/models/transformers/monkey_patch.py]: added
